### ISSUE TITLE
Not detaching properly from NumPy arrays used as outpus

### ISSUE THREAD

**Issue by manopapad** [nv-legate/cunumeric#344]
The following program:

```
import cunumeric as cn
import numpy as np
x = cn.ones((4,4))
y = cn.ones((4,4))
z = np.zeros((4,4)) # note: NumPy array
cn.dot(x, y, out=z)
```

crashes at shutdown with:

```
[0 - 1116d9dc0]    1.384084 {6}{event}: Event 8000000002a00000 destroyed with current local waiters
[0 - 1116d9dc0]    1.384118 {6}{event}:   waiting on 8000000002a00010: 0x7fe902473328
```

Here is what I see when debugging this:

```
[0 - 10029fdc0]    4.123770 {6}{event}: Event 8000000002600000 destroyed with current local waiters
(lldb) p *((Realm::RegionInstanceImpl::DeferredDestroy *)ew)
(Realm::RegionInstanceImpl::DeferredDestroy) $0 = {
  Realm::EventWaiter = {
    ew_list_link = {
      next = 0x0000000000000000
    }
  }
  inst = 0x000000011052e360
  mem = 0x000000011061dc60
}
(lldb) p/x $0.inst->me
(Realm::RegionInstance) $1 = (id = 0x4000000000400000)
(lldb) p/x make_event(generation.load())
(Realm::Event) $4 = (id = 0x8000000002600008)
```

```
~> grep 'realm\|4000000000400000\|8000000002600000\|8000000002600008\|8000000002600009' /Users/mpapadakis/Sites/local/2022/05/09/172434/legate_0.log
[0 - 700008922000]    3.942487 {2}{inst}: creating new local instance: 4000000000400000
[0 - 700008922000]    3.942646 {2}{inst}: instance created: inst=4000000000400000 external=memory(base=1104c11b0, size=128) ready=0
[0 - 70000851a000]    4.021355 {2}{event}: user event created: event=8000000002600008
[0 - 70000851a000]    4.021380 {2}{event}: user event trigger: event=8000000002600008 wait_on=0
[0 - 700008316000]    4.022003 {2}{event}: user event created: event=8000000002600009
[0 - 700008316000]    4.022012 {2}{inst}: instance destroyed: inst=4000000000400000 wait_on=8000000002600009
[0 - 70000851a000]    4.069760 {2}{realm}: shutdown requested - wait_on=8000000000400005 code=0
[0 - 70000851a000]    4.073912 {2}{realm}: triggering deferred shutdown
[0 - 10029fdc0]    4.073955 {2}{realm}: shutdown request received - terminating
[0 - 10029fdc0]    4.073964 {2}{realm}: local processor shutdown tasks initiated
[0 - 10029fdc0]    4.074792 {2}{realm}: local processor shutdown tasks complete
[0 - 10029fdc0]    4.123742 {6}{event}: Event 8000000002600000 destroyed with current local waiters
```

If I'm interpreting this correctly, instance `4000000000400000` is attached to the memory of the NumPy array, and its destruction is tied to user event `8000000002600009`, which is never triggered.

Note that this doesn't happen if we simply use NumPy arrays for inputs.

**Comment by lightsighter** [comment]
So there are really two "bugs" here:

1. cuNumeric is leaking a reference to something and keeping the instance around. It's unclear whether this is desired behavior or not.
2. There's definitely a bug in the "force deletion" routine that Legion has to prevent users seeing this particular error from Realm.

I'd like to at least fix #2, but at the moment I can't reproduce this error with what I think are the most recent versions of the core (b3e9ae1cc1db3) and cuNumeric (1b37d8b02dc9f). Is there something special I need to do to get this to trigger?
